src/Database/Esqueleto/PostgreSQL/JSON.hs

```diff
@@ -0,0 +1,618 @@
+(--.) :: SqlExpr (Value (Maybe Aeson.Value))
+      -> [Text]
+      -> SqlExpr (Value (Maybe Aeson.Value))
+(--.) value = unsafeSqlBinOp " - " value . mkTextArray
+
+-- | This operator can remove elements nested in an object.
+-- If a 'Text' is not parsable as a number when selecting in an array
+-- (even when halfway through the selection).
+-- Negative integers delete counting from the end of an array.
+-- (e.g. @-1@ being the last element, @-2@ being the second to last, etc.)
+--
+-- __CAUTION: THIS FUNCTION THROWS AN EXCEPTION WHEN USED__
+-- __ON ANYTHING OTHER THAN OBJECTS OR ARRAYS, AND WILL__
+-- __ALSO THROW WHEN TRYING TO SELECT AN ARRAY ELEMENT WITH__
+-- __A NON-INTEGER TEXT__ (cf. examples)
+--
+-- === __Objects__
+--
+-- @
+-- {"a": 3.14, "b": null}        #- []        == {"a": 3.14, "b": null}
+-- {"a": 3.14, "b": null}        #- ["a"]     == {"b": null}
+-- {"a": 3.14, "b": null}        #- ["a","b"] == {"a": 3.14, "b": null}
+-- {"a": {"b":false}, "b": null} #- ["a","b"] == {"a": {}, "b": null}
+-- @
+--
+-- === __Arrays__
+--
+-- [true, {"b":null}, 5]       #- []            == [true, {"b":null}, 5]
+-- [true, {"b":null}, 5]       #- ["0"]         == [{"b":null}, 5]
+-- [true, {"b":null}, 5]       #- ["b"]         == ERROR: path element at position 1 is not an integer: "b"
+-- [true, {"b":null}, 5]       #- ["1","b"]     == [true, {}, 5]
+-- [true, {"b":null}, 5]       #- ["-2","b"]    == [true, {}, 5]
+-- {"a": {"b":[false,4,null]}} #- ["a","b","2"] == {"a": {"b":[false,4]}}
+-- {"a": {"b":[false,4,null]}} #- ["a","b","c"] == ERROR: path element at position 3 is not an integer: "c"
+--
+-- === __Other values__
+--
+-- 1    #- {anything} == ERROR: cannot delete from scalar
+-- "a"  #- {anything} == ERROR: cannot delete from scalar
+-- true #- {anything} == ERROR: cannot delete from scalar
+-- null #- {anything} == ERROR: cannot delete from scalar
+-- @
+--
+-- === __PostgreSQL Documentation__
+--
+-- /Works with any PostgreSQL of version >= 9.5/
+--
+-- @
+--     | Type   | Description                                             |  Example
+-- ----+--------+---------------------------------------------------------+------------------------------------
+--  #- | text[] | Delete the field or element with specified path         | '["a", {"b":1}]'::jsonb #- '{1,b}'
+--     |        | (for JSON arrays, negative integers count from the end) |
+-- @
+--
+-- @since 3.1.0
+(#-.) :: SqlExpr (Value (Maybe Aeson.Value))
+      -> [Text]
+      -> SqlExpr (Value (Maybe Aeson.Value))
+(#-.) value = unsafeSqlBinOp " #- " value . mkTextArray
+
+mkTextArray :: [Text] -> SqlExpr (Value PersistValue)
+mkTextArray = val . PersistArray . fmap toPersistValue
+
+
+
+--- ORPHAN INSTANCES ---
+--- ORPHAN INSTANCES ---
+--- ORPHAN INSTANCES ---
+
+
+-- Mainly copied over from Database.Persist.Postgresql.JSON
+-- Since we don't need anything else and adding another dependency
+-- just for these two instances is a bit overkill.
```

Comment:
`esqueleto` already depends on `persistent-postgresql`. I'd *really* rather not have orphan instances potentially clashing and would definitely prefer to bring on another dependency to avoid that.

Reply:
The library `esqueleto:lib` doesn't.

But if you'd rather have it added in, sure, I can do that.

Reply:
Good catch :sweat_smile: 

In that case I think it would *definitely* cause problems to use an orphan instance here, as most users of the library will already have `persistent-postgresql` orphan instances in scope. But then why are the tests working? I'd expect them to pick up both instances and get confused. :thinking: So maybe it's not an issue.

A consequence of adding `persistent-postgresql` as a dependency to `esqueleto` is that everyone who builds the library will need all the Postgres stuff that `pkg-config` expects for it. That's not good - mysql/sqlite users shouldn't have to install this.

It may be time to switch the DB specific modules to DB specific packages. `esqueleto-postgresql` can 100% depend on `persistent-postgresql`, and then none of this is a problem. I think that's probably the "Right" answer, but it's also the most work, so .. . .

Reply:
The reason the tests work is because the instances for `Value` are only exported from `Database.Persist.Postgresql.JSON`, and that isn't imported anywhere.

Also, the ONLY reason the test has a `persistent-postgresql` dependency, is that it uses `withPostgreslConn`. Which, I expect, is fairly easily replaced by using stuff from the `postgresql-simple` library.

I tried, and yes, importing both `Database.(Persist|Esqueleto).PostgreSQL.JSON` modules results in compiler errors...
It is a predicament, because we can't put the Aeson.Value instances in the `persistent` library, since they are specific to PostgreSQL...

We could just add a notice: __"Don't use both these methods in the same module"__?

Reply:
No, that's not acceptable, because the orphans propagate. Suppose I had two modules:

```
module A where 

import Database.Esqueleto.Postgresql.JSON ()
```

```
module B where

import Database.Persist.Postgresql.JSON ()
```

`module B` and `module A` both export the orphan instance implicitly now. So any module that imports `B` and `A` and tries to use a function depending on the `instance PersistField Value` will give an error.

Reply:
That's not actually right, sorry. If I define `a :: SqlPersistM Value` and `b :: SqlPersistM Value` that use the `PersistField` instances internally, then it will select the instance that was imported for those modules. Using those functions in a `module C` that imported both would be fine.

However, if we try to use the instance in any module which imports both `B` and `A` transitively, then we will get the error due to instance overlap.

Reply:
Ah, now I get why people are so against `orphan instances`.

So yeah, splitting the modules into their own packages would make the most sense, probably. Or at least to split off the `PostgreSQL.*` parts.

Reply:
I need to get some broader input on what's the best release policy for this. It's a bit tricky and I'm not sure the best way to go about it.

Removing the `.Postgresql` and `.Mysql` modules is a breaking change and would signal a major bump. We can't just reexport from those modules because then we incur the dependency that we're trying to avoid. Releasing an `esqueleto-core` package without the "extras" and keeping `esqueleto` as the "kitchen sink" might work out without breaking stuff.

Reply:
Maybe the use of `Value` can be hidden behind a `newtype` internally so the instance doesn't leak. But that'd allow the instance to drift between `persistent-postgresql` and `esqueleto`.

Reply:
Would that make it incompatible for tables defined with Aeson.Value (for `persistent-postgresql`) to be used by the `esqueleto` functions? I mean, you could make a coercion function `SqlExpr (Value Aeson.Value) -> SqlExpr (Value NewType)`, but the other way around (defining the table with the `NewType`) would make it impossible to use the `persistent-postgresql` functions... is that an ok trade-off? Make it clear to the user that this issue exists and that, for interoperability between `persistent-postgresql` and `esqueleto`, you should define your tables with `Aeson.Value` and not the `NewType`?

Reply:
I think I might have the solution. Adding the orphan instances as their separate module makes them optional to import. The `Database.Esqueleto.PostgreSQL.JSON` module doesn't need the instances, only when using them in queries. So if you don't have `persistent-postgresql` you can import the `Database.Esqueleto.PostgreSQL.JSON.Instances` module, and otherwise, just use the one from `persistent-postgresql`.

It seems to work, I tested it.

Reply:
Implemented JSONB newtype for `PersistField(Sql)` instances and also type polymorphism, as long as the provided type has `From-/ToJSON` instances.